### DISCUSSION TITLE
The threadJoin function and finalize function of the UserObject class

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26171] [Q&A General] [answered]
I'm planning to write a subclass of DomainUserObject to implement integration calculations for variables over all faces of a mesh. For reference, I've examined two classes, NodalArea and ElementSidesL2Norm. According to the MOOSE website, the NodalArea object computes the tributary area of each node on a contact surface, while ElementSidesL2Norm computes the L2 norm of a variable over element sides. However, I'm curious about how MOOSE implements the traversal of these classes on the mesh.

They are subclasses of InternalSideUserObject and SideUserObject, respectively, and both of these base classes have references to Elem and Side that are const-qualified:
```
/// pointer to the current element object
  const Elem * const & _current_elem;

/// current side of the current element
  const unsigned int & _current_side;
```
From the MOOSE code, I suspect that the threadJoin and finalize functions handle the information transfer and processing for calculations over all element faces. All subclasses of UserObject (including NodalArea and ElementSidesL2Norm) have these two functions. In UserObject.h, their code comments are as follows:
```
  /**
   * Finalize.  This is called _after_ execute() and _after_ threadJoin()!  This is probably where
   * you want to do MPI communication!
   */
  virtual void finalize() = 0;
  /**


   * Must override.
   *
   * @param uo The UserObject to be joined into _this_ object.  Take the data from the uo object and
   * "add" it into the data for this object.
   */
  virtual void threadJoin(const UserObject & uo) = 0;
```
ElementSidesL2Norm inherits from InternalSideIntegralPostprocessor, and in InternalSideIntegralPostprocessor, MOOSE defines the threadJoin and finalize functions:
```
void
InternalSideIntegralPostprocessor::threadJoin(const UserObject & y)
{
  const InternalSideIntegralPostprocessor & pps =
      static_cast<const InternalSideIntegralPostprocessor &>(y);
  _integral_value += pps._integral_value;
}

void
InternalSideIntegralPostprocessor::finalize()
{
  gatherSum(_integral_value);
}
```
For NodalArea, the definitions of these two functions are as follows:
```
void
NodalArea::threadJoin(const UserObject & fred)
{
  const NodalArea & na = dynamic_cast<const NodalArea &>(fred);

  std::map<const Node *, Real>::const_iterator it = na._node_areas.begin();
  const std::map<const Node *, Real>::const_iterator it_end = na._node_areas.end();
  for (; it != it_end; ++it)
  {
    _node_areas[it->first] += it->second;
  }
}

void
NodalArea::finalize()
{

  const std::map<const Node *, Real>::iterator it_end = _node_areas.end();
  for (std::map<const Node *, Real>::iterator it = _node_areas.begin(); it != it_end; ++it)
  {
    const Node * const node = it->first;
    dof_id_type dof = node->dof_number(_system.number(), _variable->number(), 0);
    _aux_solution.set(dof, 0);
  }
  _aux_solution.close();

  for (std::map<const Node *, Real>::iterator it = _node_areas.begin(); it != it_end; ++it)
  {
    const Node * const node = it->first;
    dof_id_type dof = node->dof_number(_system.number(), _variable->number(), 0);
    _aux_solution.add(dof, it->second);
  }
  _aux_solution.close();
}
```
Regarding the definition in NodalArea, I find it quite perplexing. While I can understand the accumulation of _integral_value in ElementSidesL2Norm, why is there an accumulation of _node_areas in the threadJoin function of NodalArea? Additionally, the current NodalArea object seems to possess nodal information on other element sides. Since threadJoin already facilitates information exchange between different UserObjects, why is the finalize function necessary?

The implementation of threadJoin and finalize functions gives me the impression that objects declared using ElementSidesL2Norm and NodalArea don't just have information from one element side but rather seem to hold the computation results for all mesh elements. This issue has been bothering me for quite some time, and I would greatly appreciate any guidance if someone could provide some insight!!!

**Comment by GiudGiud** [comment, marked as answer]
Hello

>  threadJoin and finalize functions gives me the impression that objects declared using ElementSidesL2Norm and NodalArea don't just have information from one element side but rather seem to hold the computation results for all mesh elements.

Well yes they do hold computational results for all 'mesh elements' (as opposed to just a single piece of data). In different ways, as ElementSidesL2Norm computes a single number and NodalArea computes one number per node, but these threadJoin and finalize steps are meant to reduce information across threads and parallel processes respectively. 

ThreadJoin is across threads. Any thread can have computed any part of the nodal area of any node. 
Finalize reduces across MPI processes. Any MPI process can have computed any part of the nodal area of any node. 
For both of these we need reduce across the workers to only keep one number per-node

Guillaume

**Reply by hityyds** [reply]
Thank you very much for your response. I have received your assistance several times recently. Regarding your reply, I have the following understanding; I'm not sure if my interpretation is correct.

> Well yes they do hold computational results for all 'mesh elements' (as opposed to just a single piece of data).

Does this imply that there are NodalArea and ElementSideL2Norm objects, each associated with a face on a mesh cell, and even though each object contains information for only one face, it ultimately stores the computation results for all cells? Could this potentially lead to significant memory wastage since the results for all faces are duplicated multiple times?

> these threadJoin and finalize steps are meant to reduce information across threads and parallel processes respectively.

In terms of the execution logic, does MOOSE initially consolidate the computation results for all threads within each process by invoking the threadJoin function, and subsequently call the finalize function to aggregate the computation results for all processes? This would ensure that each NodalArea and ElementSideL2Norm object holds the complete computation results for the entire mesh.

**Reply by GiudGiud** [reply]
To the first one, yes and no

We only have one copy of the object doing the work for every thread and every MPI process. We don't need one object per element, each object can process multiple elements.

However they each hold data that will need to be reduced at the end (after they have been accumulated on the thread/process object, from the elements/nodes that that copy examined)

For the NodalArea that data is on a per node basis. So yes it can be fairly large and replicating it will cause memory issues. However, it is not of the size of the entire mesh's number of nodes. It s only sized to the nodes that this object is actually looking at. So we do not expect a total replication. 

When you run in this kind of memory concerns, you may consider distributing the mesh. This will ensure process-storage remains local as each process will only see part of the mesh

**Reply by GiudGiud** [reply]
To the second one yes.

TheeadJoin first to reduce across threads then finalize to reduce across processes. 

I think these are global all-to-all reductions (for processes at least) though it would be good to check. This means that at the end, the entire-mesh reduction is stored on every process.

**Reply by hityyds** [reply]
Therefore, if the program is running on three processes, each invoking two threads, there will be six copies. You used "reduce" to describe the functionality of the TheadJoin and finalize functions. Will these six copies persist even after executing the TheadJoin and finalize functions?

> They are subclasses of InternalSideUserObject and SideUserObject, respectively, and both of these base classes have references to Elem and Side that are const-qualified:
> ```
> /// pointer to the current element object
>   const Elem * const & _current_elem;
> 
> /// current side of the current element
>   const unsigned int & _current_side;
> ```

Furthermore, as I mentioned in the question, these objects have references to elem and side that are const-qualified. In this context, how do they handle computations involving multiple elems? I haven't undergone formal programming education, but based on my understanding, this referencing approach doesn't seem to enable an object to perform calculations on multiple elems because it only owns data related to one elem side.

**Reply by GiudGiud** [reply]
> Therefore, if the program is running on three processes, each invoking two threads, there will be six **copies.**

yes

> Will these six copies persist even after executing the TheadJoin and finalize functions?

yes, for the next iteration

> Furthermore, as I mentioned in the question, these objects have references to elem and side that are const-qualified. In this context, how do they handle computations involving multiple elems? 

the reference is const, but the thing it references can change under the hood. That is done by the Assembly class usually. Here I think it's done by the Threaded side-element loop. The 'constness' is only local to the object, it means that NodalArea cannot modify that reference, not that that reference cannot ever change from somewhere else

**Reply by hityyds** [reply]
> ```
> /// pointer to the current element object
>   const Elem * const & _current_elem;
> ```

Is this not declaring a constant pointer to a constant element? I always thought that neither the pointer should point to a new address nor should the value being pointed to be changed.

> Here I think it's done by the Threaded side-element loop. 

I suspect that MOOSE performs the mesh traversal through the loop in ThreadedElementLoopBase::operator()(const RangeType & range, bool bypass_threading), which repeatedly calls ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side) and other UserObjects' computation functions. However, I couldn't find in ComputeUserObjectsThread::onInternalSide() how MOOSE invokes the relevant elem and side for the computation based on the input parameters ('elem' and 'side'). Instead, it seems MOOSE repeatedly calls a batch of InternalSideUserObject objects without changing elem and side.
```
void
ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
{
  // Pointer to the neighbor we are currently working on.
  const Elem * neighbor = elem->neighbor_ptr(side);

  // Get the global id of the element and the neighbor
  const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();

  if (_internal_side_objs.size() == 0 && _domain_objs.size() == 0)
    return;
  if (!((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) ||
        (neighbor->level() < elem->level())))
    return;

  _fe_problem.prepareFace(elem, _tid);
  _fe_problem.reinitNeighbor(elem, side, _tid);

  // Set up Sentinels so that, even if one of the reinitMaterialsXXX() calls throws, we
  // still remember to swap back during stack unwinding.
  SwapBackSentinel face_sentinel(_fe_problem, &FEProblem::swapBackMaterialsFace, _tid);
  _fe_problem.reinitMaterialsFace(elem->subdomain_id(), _tid);

  SwapBackSentinel neighbor_sentinel(_fe_problem, &FEProblem::swapBackMaterialsNeighbor, _tid);
  _fe_problem.reinitMaterialsNeighbor(neighbor->subdomain_id(), _tid);

  for (const auto & uo : _internal_side_objs)
    if (!uo->blockRestricted() || uo->hasBlocks(neighbor->subdomain_id()))
      uo->execute();

  for (auto & uo : _domain_objs)
    if (!uo->blockRestricted() || uo->hasBlocks(neighbor->subdomain_id()))
    {
      uo->preExecuteOnInternalSide();
      uo->executeOnInternalSide();
    }
}
```

**Reply by GiudGiud** [reply]
> const Elem * const & _current_elem;

this is a reference to a constant pointer to a const element (either that or a const reference to an element pointer, not sure)
1st const says that NodalArea does not modify this element
2nd const says that NodalArea does not modify this pointer (once it is set)

The important thing is that the element can change! Just not changed by NodalArea

**Reply by hityyds** [reply]
Thank you very much for your response!
